src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/Completion.cs

```diff
@@ -271,4 +282,135 @@ private static TextEdit BuildRevertedEdit(TextEdit provisionalTextEdit)
         var formattingOptions = _formattingOptionsProvider.GetOptions(document.TextDocumentIdentifier.Uri);
         return Task.FromResult(formattingOptions);
     }
+
+    [JsonRpcMethod(LanguageServerConstants.RazorSnippetCompletionEndpointName, UseSingleObjectParameterDeserialization = true)]
+    public CompletionItem[]? GetSnippetCompletions(RazorSnippetCompletionParams completionParams, CancellationToken _)
+    {
+        if (completionParams.TextSpan.Length == 0)
+        {
+            return null;
+        }
+
+        var snippets = _snippetCache.GetSnippets();
+
+        if (snippets.IsDefaultOrEmpty)
+        {
+            return null;
+        }
+
+        var langSpecificCompletionsEnum = completionParams.LanguageKind switch
+        {
+            RazorLanguageKind.Html => snippets.Where(static s => s.Language == SnippetLanguage.Html),
+            RazorLanguageKind.CSharp => snippets.Where(static s => s.Language == SnippetLanguage.CSharp),
```

Comment:
Perhaps `SnippetsCache` should expose two cached sets, rather than recomputing this each time?

Reply:
I like this idea. Will add in next change :)